### ISSUE TITLE
less.refresh() promise fires back before its done

### ISSUE THREAD

**Issue by arapoport** [less/less.js#2844]
Hi,

I think I found a bug in less.js. I have a web application that loads around ten less files and compiles them in browser. I trigger their compilation using less.refresh(). But the promise returned by this function sometimes fires before all styles were compiled and applied.

If I have something like this:
`less.refresh().done(function(){ console.log("less should be done by now"); })`

It may happen that after my log line gets printed further less files are compiled and their log gets printed (like "css for foo.less generated in 59ms")

It looks like it waits for the last file (by order) to be ready and then fires the promise. However sometimes the last file may be ready before others.

A patch would be to fix the function less.refresh as follows (note the variable remainingSheets):

```
less.refresh = function (reload, modifyVars, clearFileCache) {
        if ((reload || clearFileCache) && clearFileCache !== false) {
            fileManager.clearFileCache();
        }
        return new Promise(function (resolve, reject) {
            var startTime, endTime, totalMilliseconds;
            startTime = endTime = new Date();       

            // Change 1
            var remainingSheets = less.sheets.length;
           
            loadStyleSheets(function (e, css, _, sheet, webInfo) {
                if (e) {
                    errors.add(e, e.href || sheet.href);
                    reject(e);
                    return;
                }
                if (webInfo.local) {
                    less.logger.info("loading " + sheet.href + " from cache.");
                } else {
                    less.logger.info("rendered " + sheet.href + " successfully.");
                }
                browser.createCSS(window.document, css, sheet);
                less.logger.info("css for " + sheet.href + " generated in " + (new Date() - endTime) + 'ms');

                // Change 2
                remainingSheets--;
                // Change 3
                if (remainingSheets === 0) { // (webInfo.remaining === 0) {
                    totalMilliseconds = new Date() - startTime;
                    less.logger.info("less has finished. css generated in " + totalMilliseconds + 'ms');
                    resolve({
                        startTime: startTime,
                        endTime: endTime,
                        totalMilliseconds: totalMilliseconds,
                        sheets: less.sheets.length
                    });
                }
                endTime = new Date();
            }, reload, modifyVars);

            loadStyles(modifyVars);
        });
    };
```

Let me know if you want me to submit a proper patch.

**Comment by matthew-dean** [comment]
It would be great if you could submit a patch. Also, there is an edge case where if you are using inline styles or there are no stylesheets, the promise never fires, which shouldn't happen. (#2812)

**Comment by arapoport** [comment]
Hi, yes I can :-) 

I already fixed it locally and I am about to submit it. Original tests work well. Should I add tests for this feature? If yes, can you tell me the place? I am not sure how to write them...

**Comment by seven-phases-max** [comment]
Closing as fixed in #2851.
